### ISSUE TITLE
Allow multi selection of entries in QTreeView widgets (Functions, Strings, etc.) 

### ISSUE THREAD

**Issue by ITAYC0HEN** [radareorg/cutter#766]
Currently, multi-selection of items in our widgets isn't possible (at least in those I checked).
It would be great to allow the user to select multiple items so he can count them (after #754 would be fixed) and/or copy them.

1. clicking on entry A and while holding Shift a click on entry B will select all items in the range between A to B
2. Holding CTRL and then clicking on an item will select it if it is not selected already and deselect it if it is. (aka toggle)
3. Using the mouse to select multiple items by holding the left button and moving the cursor
4. Ctrl+C will copy all the selected rows

**Comment by duck2** [comment]
I can give this a try.

**Comment by ITAYC0HEN** [comment]
It would be great @duck2! Thank you :)
Please update us with the progress

**Comment by duck2** [comment]
Hello,

After doing some research, I think I should first subclass `QTreeView` into a custom class, say, `CutterQTreeView` and implement the functionality using [QItemSelectionModel](http://doc.qt.io/qt-5/qitemselectionmodel.html).

This will require changing code in all widget files(`QTreeView to CutterQTreeView` and `<customwidget>` tags). Is that OK?

**Comment by xarkes** [comment]
It sounds ok, but: I think someone is already creating a subclass `CutterQTreeView` (https://github.com/radareorg/cutter/pull/779)
So it would be better if one could wait on the other (I let you decide between yourselves) to try not to do the same work at the same time.
cc @smhaziq

**Comment by duck2** [comment]
I can wait until the subclass is created, given I didn't get Cutter to build with `CutterQTreeView` yet.

**Comment by xarkes** [comment]
@duck2 The PR #779 has been merged, so feel free to extend the class.

**Comment by ITAYC0HEN** [comment]
ping @duck2

**Comment by duck2** [comment]
Pong, I now know how to implement this, it's
```
FooTreeView->setSelectionMode(QAbstractItemView::ExtendedSelection);
```
I checked out the new commits, but I cannot find any way to add this to the `CutterTreeWidget` class, since I couldn't find a way to access the tree view.

The `CutterTreeWidget` is initialized before the `ui->setupUi` is called, so it shouldn't have access to parent's other children via `parent->ui->findChild<QTreeView *>`.

I think the subclass should be something which manages its own status bar(why should parents update it) and exposes(or _is_) a `TreeView` object for the parent widgets to customize.

As an alternative, I can just add the above line to every widget which has a `QTreeView`, but it's not elegant.

Perhaps I'm overthinking this. Open to any redirections.

**Comment by smhaziq** [comment]
> I think the subclass should be something which manages its own status bar(why should parents update it) and exposes(or is) a TreeView object for the parent widgets to customize.

Agreed. But as I said in #779 , I didn't put much thought on the CutterTreeWidget class design as it's an afterthought over a small task so I hope it get refined here. Feel free to change the class design as long the status bar still correctly get updated.

But if you don't want to make major change, how about passing the TreeView to the CutterTreeWidget class and process the view from there?
```
tree->setView(ui->flagsTreeView);
```

**Comment by ITAYC0HEN** [comment]
ping @duck2 @smhaziq . Can you share the progress of handling this issue?
we would be happy to help and direct :)

**Comment by smhaziq** [comment]
@duck2 are you still working on this?

I think your suggestion to subclass the Qtreeview is correct as we need to overwrite the copy behavior as well so it support rows copying.

And to subclass it with current CutterTreeWidget class implementation is wrong since it deals with the widget's layout as well for statusbar control.

Maybe CutterTreeWidget should take over the ui creation (replacing .ui) as well and from there we can subclass the Qtreeview properly.

**Comment by duck2** [comment]
Hello,

I _postponed_ this, and now the school fully started. I do want to deliver though.

Opening a PR to show where am I stuck.

**Comment by xarkes** [comment]
Fixed in #903
